pypesto/visualize/sampling.py

```diff
@@ -91,6 +114,565 @@ def sampling_fval_trace(
+        labels: Dict[str, str] = None,
+        axis_label_padding: int = 30,
+        groupby: str = CONDITION,
+        condition_gap: float = 0.01,
+) -> matplotlib.axes.Axes:
+    """Plot MCMC-based prediction confidence intervals for the
+    model states or observables. One or various confidence levels
+    can be depicted. Plots are grouped by condition.
+
+    Note that references to observables here corresponds to the meaning of an
+    observable in the pyPESTO predictions code, which means an observable could
+    refer to many things, including model states or observables, dependent on
+    how the prediction is processed. FIXME rename
+
+    Parameters
+    ----------
+    result:
+        The pyPESTO result object with filled sample result.
+    levels:
+        Credibility levels, e.g. [95] for a 95% credibility interval. See the
+        :py:func:`_get_level_percentiles` method for a description of how these
+        levels are handled, and current limitations.
+    title:
+        Axes title.
+    size: ndarray
+        Figure size in inches.
+    axes:
+        Axes object to use.
+    labels:
+        Keys should be ensemble observable IDs, values should be the desired
+        label for that observable. Defaults to observable IDs.
+    axis_label_padding:
+        Pixels between axis labels and plots.
+    groupby:
+        Group plots by `OBSERVABLE` or `CONDITION`. TODO single constants file?
+    condition_gap:
+        Gap between conditions when `groupby == CONDITION`.
+
+    Returns
+    -------
+    axes:
+        The plot axes.
+    """
+    if labels is None:
+        labels = {}
+    if len(list(levels)) == 1:
+        levels = list(levels)
+    levels = sorted(levels, reverse=True)
+    percentiles = [
+        percentile
+        for level in levels
+        for percentile in _get_level_percentiles(level)
+    ]
+
+    ensemble_prediction.compute_summary(percentiles_list=percentiles)
+
+    summary = ensemble_prediction.prediction_summary
+
+    all_condition_ids = [
+        prediction.condition_ids
+        for prediction in ensemble_prediction.prediction_summary.values()
+    ]
+    # All prediction results must predict for the same set of conditions.
+    # Can support different conditions later.
+    if not (
+            np.array([
+                set(condition_ids) == set(all_condition_ids[0])
+                for condition_ids in all_condition_ids
+            ]).all()
+    ):
+        raise KeyError('All predictions must have the same set of conditions.')
+    condition_ids = all_condition_ids[0]
+
+    observable_ids = sorted({
+        observable_id
+        for prediction in ensemble_prediction.prediction_summary.values()
+        for condition in prediction.conditions
+        for observable_id in condition.observable_ids
+    })
+
+    # Set default labels.
+    labels = {
+        k: (labels[k] if k in labels else k)
+        for k in condition_ids + observable_ids
+    }
```

Comment:
The if "labels is None" would better fit here

Reply:
Hm, the current code is such that only specifying a subset of labels is OK, e.g. equivalent to
```python
default_labels = dict(...)
my_custom_labels = dict(...)
labels = {**default_labels, **my_custom_labels}
```

Refactored to the following, hopefully more readable
```python
labels = {id_: labels.get(id_, id_) for id_ in condition_ids + output_ids}
```